ugs-core/src/com/willwinder/universalgcodesender/connection/TCPConnection.java

```diff
@@ -0,0 +1,165 @@
+	private InputStreamReader inStream;
+	private Thread replyThread;
+	private ResponseMessageHandler responseMessageHandler;
+
+	@Override
+	public void setUri(String uri) {
+		try {
+			host = StringUtils.substringBetween(uri, ConnectionDriver.TCP.getProtocol(), ":");
+			port = Integer.valueOf(StringUtils.substringAfterLast(uri, ":"));
+		} catch (Exception e) {
+			throw new ConnectionException("Couldn't parse connection string " + uri, e);
+		}
+
+		if (StringUtils.isEmpty(host)) {
+			throw new ConnectionException("Empty host in connection string.");
+		}
+		if ((port < 1) || (port > 65535)) {
+			throw new ConnectionException("Please ensure port is a port number between 1 and 65535.");
+		}
+	}
+
+	@Override
+	public boolean openPort() throws Exception {
+		responseMessageHandler = new ResponseMessageHandler();
+
+		try {
+			client = new Socket(host, port);
+		} catch( BindException e) {
+			throw new ConnectionException("Could not bind a local port.");
+		} catch( NoRouteToHostException e) {
+			throw new ConnectionException("No route to host. The remote host may not be running, blocked by a firewall, or disconnected.");
+		} catch( PortUnreachableException e) {
+			throw new ConnectionException("The port is unreachable on the remote host. The server may not be running, or blocked by a firewall.");
+		} catch (SocketException e) {
+			e.printStackTrace();
+		}
+		if (client == null) {
+			throw new ConnectionException("Socket unable to connect.");
+		}
+
+		bufOut = client.getOutputStream();
+		inStream = new InputStreamReader(client.getInputStream());
+		bufIn = new BufferedReader(inStream);
+
+		// start thread so replies can be handled
+		replyThread = new Thread(this);
+		replyThread.start();
+
+		return client.isConnected();
+	}
+
+	@Override
+	public void closePort() throws Exception {
+		if (client != null) {
+			try {
+				replyThread.interrupt();
+				client.close();
+			} catch (SocketException e) {
+				// ignore socketexception if connection was broken early
+			} finally {
+				client = null;
+			}
+		}
+	}
+
+	@Override
+	public boolean isOpen() {
+		return (client != null) && (!client.isClosed());
+	}
+
+	/**
+	 * Sends a command to remote host.
+	 * @param command Command to be sent to remote host.
+	 */
+	public void sendStringToComm(String command) throws Exception {
+		bufOut.write(command.getBytes());
+		bufOut.flush();
+	}
+
+	/**
+	 * Immediately sends a byte, used for real-time commands.
+	 */
+	public void sendByteImmediately(byte b) throws Exception {
+		bufOut.write(b);
+		bufOut.flush();
+	}
+
+	/**
+	 * Thread to accept data from remote host, and pass it to responseHandler
+	 */
+	public void run() {
+		String resp;
+		while(!Thread.interrupted() && !client.isClosed())
+		{
+			try {
+				if(inStream.ready() && (resp = bufIn.readLine()) != null) {
+					responseMessageHandler.handleResponse(resp + "\n", comm);
+				}
+			} catch (SocketException e) {
+				e.printStackTrace();
+				System.exit(-1);
```

Comment:
This should probably be handled in some other way than just closing the software (is this what happens if an error occur in a separate thread in Netbeans?)